lib/io/stdin.fz

```diff
@@ -35,14 +35,46 @@ private stdin(
   ) : stdins, effect r is
 
 
-  # for a stdin instance installed in the current environment,
-  # install next instance of byteInputHandler,
-  # 'stdin.env' will provide the new instance.
-  #
-  private next => stdin ip.next mode unit
-
-  nextByte =>
-    next.ip.get
+  # read a single byte from stdin
+  byte =>
+    (stdin ip.next mode unit).ip.get
+
+
+  # an infinite list of either a byte, an end_of_file or an error
+  # Caution!: this list is not immutable since it is constructed lazily by reading from stdin!
+  # you may iterate only once.
+  # Additionally there may be read more from stdin than what is consumed
+  # via this list.
+  # NYI: How can we lessen these imperfections?
+  bytes list (choice u8 end_of_file error) is
+    list byte ()->bytes
+
+
+  # try reading a codepoint from stdin
+  codepoint =>
+    read_codepoint(read_bytes list u8) choice error end_of_file codepoint is
+      match byte
+        byte u8 =>
+          bytes := read_bytes ++ [byte]
+          match (strings.fromBytes bytes).codepointsAndErrors.first
+            c codepoint => c
+            e error => if bytes.count >= 4 then e else read_codepoint bytes
+        end_of_file => end_of_file
+        e error => e
+
+    read_codepoint (lists.empty u8)
+
+
+  # a - possibly infinite - list of codepoints read from stdin
+  # Caution!: this list is not immutable since it is constructed lazily by reading from stdin!
+  # you may iterate only once.
+  # Additionally there may be read more from stdin than what is consumed
+  # via this list.
+  # NYI: How can we lessen these imperfections?
```

Comment:
One way around this would be to memoize the result, i.e, add a field
```
codepoints_result := mut (option (list codepoint) nil)
```
and then check this in `codepoints`, return it if it is non-`nil` and read the next codepoint if it is. In case a non-empty list is returned, the new list must itself contain the `head` codepoint and use memoization for `tail`.

Reply:
@fridis I think there might be other subtle issues though.
Lets say i do something like `codepoints.take 0` I would expect it to read zero `codepoints`. But actually it would read one since it has to construct a cons cell.

Reply:
`take 0` just returns `nil`.